### PR TITLE
HDDS-6321. Avoid refresh pipeline for key lookup in checkAcls

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Reduce RPC calls by avoiding content-related operations (refresh pipeline, etc.) for ACL check.  Reuse `isHeadOp` flag for this purpose.

https://issues.apache.org/jira/browse/HDDS-6321

## How was this patch tested?

Reproduced manually on `master` in `ozonesecure` environment.

Added integration test:

```
[ERROR] Tests run: 20, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 16.828 s <<< FAILURE! - in org.apache.hadoop.ozone.om.TestKeyManagerImpl
[ERROR] testCheckAccessForFileKey  Time elapsed: 0.044 s  <<< FAILURE!
org.mockito.exceptions.verification.NeverWantedButInvoked:

storageContainerLocationProtocol.getContainerWithPipelineBatch(
    <any>
);
Never wanted here:
-> at org.apache.hadoop.ozone.om.TestKeyManagerImpl.testCheckAccessForFileKey(TestKeyManagerImpl.java:486)
But invoked here:
-> at org.apache.hadoop.ozone.om.KeyManagerImpl.refreshPipeline(KeyManagerImpl.java:507)
```

and verified the change:

```
[INFO] Tests run: 20, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 15.019 s - in org.apache.hadoop.ozone.om.TestKeyManagerImpl
```

https://github.com/adoroszlai/hadoop-ozone/runs/5220817724#step:4:2886

### CODE REVIEW

**Comment by adoroszlai** [comment]
@kuenishi Thanks for reporting the problem.  Please review if you have some time.

**Review by kuenishi — COMMENTED** [review]
Looks good to me. Reusing headOp flag is very smart!

**Comment by sodonnel** [comment]
If we don't refresh the locations, does the returned KeyInfo still have locations that were stored in OM, but they may be stale? I guess the change here is for checkAccess, where `.setHeadOp(true)` is set to true now, and it is more of an internal call, so it probably doesn't matter.

How does this change impact TrashOzoneFileSystem and OzonePrefixPathImpl as `setHeadOp` is set to true there too?. I guess I am not too familiar with the OM code, so I am just trying to understand how this change fits into improving the ACL checks.

**Comment by adoroszlai** [comment]
Thanks @sodonnel for the review.

> How does this change impact TrashOzoneFileSystem and OzonePrefixPathImpl as `setHeadOp` is set to true there too?

`OzonePrefixPathImpl` is used for prefix ACLs.

You are right, `TrashOzoneFileSystem` has nothing to do with ACL check.  I looked for similar usage of `getFileStatus` / `listStatus` where the result's location information is unused.  This class converts the return value to Hadoop's `FileStatus` ignoring location:

https://github.com/apache/ozone/blob/f0ee98a45d5371fa9595f8178eca72b1bfe37e2b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/TrashOzoneFileSystem.java#L232-L250

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 1]
```diff
@@ -1244,15 +1240,19 @@ private OzoneFileStatus getOzoneFileStatus(String volumeName,
...
-        // refreshPipeline flag check has been removed as part of
-        // https://issues.apache.org/jira/browse/HDDS-3658.
-        // Please refer this jira for more details.
-        refresh(fileKeyInfo);
-        if (sortDatanodes) {
-          sortDatanodes(clientAddress, fileKeyInfo);
+        // If operation is head, do not perform any additional steps
+        // As head operation does not need any of those details.
+        if (!args.isHeadOp()) {
+          if (args.getLatestVersionLocation()) {
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:1246]: Thanks @adoroszlai for taking care this case. 
I think, need to move latest version location outside `if (!args.isHeadOp())` block.
```
        if (args.getLatestVersionLocation()) {
          slimLocationVersion(fileKeyInfo);
        }
```

**Review by rakeshadr — COMMENTED** [review]
Thanks @adoroszlai for the contribution. Added one comment, please go through it.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Thanks @rakeshadr for the review.  I have updated this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 2]
```diff
@@ -1294,17 +1295,17 @@ private OzoneFileStatus getOzoneFileStatusFSO(String volumeName,
 
     if (fileStatus != null) {
       // if the key is a file then do refresh pipeline info in OM by asking SCM
-      if (fileStatus.isFile()) {
+      if (fileStatus.isFile() && !args.isHeadOp()) {
         OmKeyInfo fileKeyInfo = fileStatus.getKeyInfo();
-        if (latestLocationVersion) {
+        if (args.getLatestVersionLocation()) {
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:1300]: @adoroszlai, same here as well. Could you please keep only refresh() and sortDatanodes() inside `if (!args.isHeadOp())` block and move LatestVersionLocation outside.

**Review by rakeshadr — APPROVED** [review]
Thanks @adoroszlai for the updates.

+1 LGTM, pending CI report

**Comment by adoroszlai** [comment]
Thanks @kuenishi, @rakeshadr, @sodonnel for the review.

**Pull request merged**